### PR TITLE
add White Noise effect

### PR DESCRIPTION
**Pull request by katsar0v**
I create an active noise effect. Here is a demo, in combination with filter and tremolo (wav in zip):

[noise_effect_demo.zip](https://github.com/mixxxdj/mixxx/files/4880909/noise_effect_demo.zip)

image

It has a "Loudness" parameter for adjusting the noise volume

### CODE REVIEW

**Comment by Be-ing** [comment]
How about we rename this "White Noise"?

**Comment by katsar0v** [comment]
Sure, it is actually called white noise, yes, will commit

File: src/effects/builtin/builtinbackend.cpp
Comment (Be-ing) [inline src/effects/builtin/builtinbackend.cpp:25]: Could you rename the file to whitenoiseeffect.h/cpp?

Comment (katsar0v) [reply src/effects/builtin/builtinbackend.cpp:25]: Okay

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.cpp:12]: Please change the class name to WhiteNoiseEffect too

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Holzhaus) [inline src/effects/builtin/whitenoiseeffect.cpp:61]: Please use static_cast instead of c-style typecasts.

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Holzhaus) [inline src/effects/builtin/whitenoiseeffect.cpp:51]: I don't think that comment is needed here. Also, we don't use block comments in our code anymore, because they break if you want to comment out a larger block of code quickly.

File: src/effects/builtin/whitenoiseeffect.h
Comment (Holzhaus) [inline src/effects/builtin/whitenoiseeffect.h]: #pragma once

File: src/effects/builtin/whitenoiseeffect.h
Comment (Holzhaus) [inline src/effects/builtin/whitenoiseeffect.h:48]: Missing newline at EOF.

File: src/effects/builtin/whitenoiseeffect.h
Comment (Holzhaus) [inline src/effects/builtin/whitenoiseeffect.h:4]: Unused include

File: src/effects/builtin/whitenoiseeffect.h
Comment (Holzhaus) [inline src/effects/builtin/whitenoiseeffect.h:24]: What is this used for?

**Review by Holzhaus — CHANGES_REQUESTED** [review]

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Holzhaus) [inline src/effects/builtin/whitenoiseeffect.cpp:62]: Please remove

**Review by Holzhaus — CHANGES_REQUESTED** [review]

File: src/effects/builtin/whitenoiseeffect.h
Comment (katsar0v) [reply src/effects/builtin/whitenoiseeffect.h:24]: This effect is forked from metronome and I forgot it, will remove it

File: src/effects/builtin/whitenoiseeffect.h
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.h:13]: WhiteNoiseGroupState

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.cpp:60]: I think this should be a dry/wet parameter. By not including the input signal at all, the only way to use this as part of a chain is at the beginning of the chain.

Also, this should be a ramping value. Otherwise, there are stairstep jumps in the volume between cycles of the audio thread instead of continuous changes. You can look at the echo effect for an example of how to implement this.

Comment (katsar0v) [reply src/effects/builtin/whitenoiseeffect.cpp:60]: I have no idea how to build what you said and maybe some assistance would be good. I saw that the echo effect has 

```
RampingValue<CSAMPLE_GAIN> send(send_current, gs.prev_send,
                                    bufferParameters.framesPerBuffer());
```

For making the ramping value and

```
CSAMPLE_GAIN send_ramped = send.getNext();
```

In the for loop

Comment (Be-ing) [reply src/effects/builtin/whitenoiseeffect.cpp:60]: Yes, you need something like that. You need to store the value of the parameter in the GroupState. That is what `gs.prev_send` is for there.

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.cpp:61]:
```suggestion
        pOutput[i] = pInput[i] * (1 - Loudness) + output;
```

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Holzhaus) [inline src/effects/builtin/whitenoiseeffect.cpp:8]:
```suggestion
    return QStringLiteral("org.mixxx.effects.whitenoise");
```

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Holzhaus) [inline src/effects/builtin/whitenoiseeffect.cpp:56]: Please don't capitalize variable names, this looks like a class name.

```suggestion
    CSAMPLE loudness = m_pLoudnessParameter->value();
```

File: src/effects/builtin/whitenoiseeffect.h
Comment (Holzhaus) [inline src/effects/builtin/whitenoiseeffect.h:2]: again: \#pragma once

File: src/effects/builtin/whitenoiseeffect.h
Comment (Holzhaus) [inline src/effects/builtin/whitenoiseeffect.h:33]: This should be wrapped. Aren't you using our [pre-commit hooks](https://github.com/mixxxdj/mixxx/wiki/Using-Git#set-up-automatic-code-checking)?

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Holzhaus) [inline src/effects/builtin/whitenoiseeffect.cpp:3]: Is this still needed?

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Holzhaus) [inline src/effects/builtin/whitenoiseeffect.cpp:1]: I think we always use includes relative to our `src/` directory:

```suggestion
#include "effects/builtin/whitenoiseeffect.h"
```

**Review by Holzhaus — CHANGES_REQUESTED** [review]

File: src/effects/builtin/whitenoiseeffect.h
Comment (katsar0v) [reply src/effects/builtin/whitenoiseeffect.h:2]: What is the solution? I saw other effects do the same

File: src/effects/builtin/whitenoiseeffect.h
Comment (katsar0v) [reply src/effects/builtin/whitenoiseeffect.h:33]: I will look into that

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (katsar0v) [reply src/effects/builtin/whitenoiseeffect.cpp:3]: QgDebug, not anymore. Thanks!

File: src/effects/builtin/whitenoiseeffect.h
Comment (crisclacerda) [reply src/effects/builtin/whitenoiseeffect.h:2]: https://en.wikipedia.org/wiki/Include_guard
#pragma once is easier and less error-prone but wasn't universally supported until recently, so mixxx codebase is still full of the "old" format. 
New code should use #pragma once

Comment (katsar0v) [reply src/effects/builtin/whitenoiseeffect.h:2]: Thank you. I will implement that. The build is failing, but from the errors I see this is not from this PR, is that correct?

File: src/effects/builtin/whitenoiseeffect.h
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.h:24]: All this needs is one `CSAMPLE_GAIN`

Comment (katsar0v) [reply src/effects/builtin/whitenoiseeffect.h:24]: Working on it

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.cpp:23]: How about renaming this to "Mix" considering it is now a dry/wet knob?

**Comment by katsar0v** [comment]
Here is a new screenshot of the White Noise effect

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (katsar0v) [reply src/effects/builtin/whitenoiseeffect.cpp:23]: I thought just naming it "Dry/Wet"

Comment (Be-ing) [reply src/effects/builtin/whitenoiseeffect.cpp:23]: I think that works too if it fits on screen without the text getting cut off.

Comment (katsar0v) [reply src/effects/builtin/whitenoiseeffect.cpp:23]: It is basically the knob that controls how loud the noise is. It also acts as a dry/wet one. Other effects use "depth" a lot for defining how intense the effect should be

File: src/effects/builtin/whitenoiseeffect.h
Comment (Holzhaus) [reply src/effects/builtin/whitenoiseeffect.h:2]: Just replace these two lines with "#pragma once" and remove the "#endif" at the bottom.

> The build is failing, but from the errors I see this is not from this PR, is that correct?

If the build worked before and you didn't pull in changes from other branches, they have to be from this PR.

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (katsar0v) [reply src/effects/builtin/whitenoiseeffect.cpp:23]: image

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.cpp:65]:
```suggestion
        float noise = static_cast<float>(
                QRandomGenerator::global()->generateDouble() * loudness_ramped);
        pOutput[i] = pInput[i] * (1 - loudness_ramped) + noise * loudness_ramped;
```
`output` could be confused with `pOutput`

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.cpp:38]:
```suggestion
        : m_pDryWetParameter(pEffect->getParameterById("Dry/Wet")) {
```
Rename all the code variables too

File: src/effects/builtin/whitenoiseeffect.h
Comment (katsar0v) [reply src/effects/builtin/whitenoiseeffect.h]: Hm, I am not sure the build has worked before. Some of the automated tests succeed though. And it always compiles on my pc :)

Comment (Holzhaus) [reply src/effects/builtin/whitenoiseeffect.h]: I didn't realize you were talking about the CI builds. It still builds on bionic which doesn't have QRandomGenerator because the qt version is too old.

**Comment by katsar0v** [comment]
It seems that the build fails, because QRandomGenerator could not be found. It was introduced with Qt 5.10, which version does Travis use and which Qt version should be used for development? I have 5.12

**Comment by Holzhaus** [comment]
> It seems that the build fails, because QRandomGenerator could not be found. It was introduced with Qt 5.10, which version does Travis use and which Qt version should be used for development? I have 5.12

This decision is currently being discussed here: https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/minimum.20requirements.20policy

If we need to support Qt 5.10 for Mixxx 2.4, you need to make a `QT_VERSION_CHECK` that falls back `qrand` and divide the result by `RAND_MAX`. You can grep the source to see how the version check is used in other parts of the code base.

**Comment by katsar0v** [comment]
Ok, I will add this, thank you

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.cpp:56]: extra empty line

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.cpp:61]: Rename all "loudness" variables to "dryWet"

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.cpp:23]: conventionally we use lowercase for the ID strings
```suggestion
    intensity->setId("dry_wet");
```

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (katsar0v) [reply src/effects/builtin/whitenoiseeffect.cpp]: Thank you. Fixed.

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.cpp:71]:
```suggestion
        float noise = static_cast<float>(
            r_distributor(gs.gen));
        
        pOutput[i] = pInput[i] * (1 - drywet_ramped) + noise * drywet_ramped;
```
You are multiplying by `drywet_ramped` twice.

File: src/effects/builtin/whitenoiseeffect.h
Comment (katsar0v) [reply src/effects/builtin/whitenoiseeffect.h]: Yeah, this should be fine now I hope

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.cpp:23]: Can you move the `dry_wet` string to a named constant in an anonymous namespace at the top of the file? This way it doesn't need to be updated in two places if it ever changes. I intend to do this for the rest of the built in effects after #2618 is merged.

```
namespace {
constexpr QString dryWetParameterId = "dry_wet";
} // anonymous namespace
```

Comment (katsar0v) [reply src/effects/builtin/whitenoiseeffect.cpp:23]: > Can you move the `dry_wet` string to a named constant in an anonymous namespace at the top of the file? This way it doesn't need to be updated in two places if it ever changes. I intend to do this for the rest of the built in effects after #2618 is merged.
> 
> ```
> namespace {
> constexpr QString dryWetParameterId = "dry_wet";
> } // anonymous namespace
> ```
In the header file or .cpp?

Comment (Holzhaus) [reply src/effects/builtin/whitenoiseeffect.cpp:23]: I don't think QString can be made constexpr. 
```cpp
namespace {
const QString dryWetParameterId = QStringLiteral("dry_wet");
} // anonymous namespace
```

Comment (Holzhaus) [reply src/effects/builtin/whitenoiseeffect.cpp:23]: In the cpp file.

**Comment by Be-ing** [comment]
Please sign the [Mixxx Contributor Agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ) before we can merge your code. Leave a comment when you have signed.

**Comment by katsar0v** [comment]
I signed the agreement

**Comment by katsar0v** [comment]
It seems that the build fails with scons, but succeeds with cmake. Is there something I can do to fix this? It complains about undefined reference which I don't understand

**Comment by Holzhaus** [comment]
> It seems that the build fails with scons, but succeeds with cmake. Is there something I can do to fix this? It complains about undefined reference which I don't understand

You only added the file to CMakeLists.txt, not to `build/depends.py`.

**Comment by katsar0v** [comment]
Got it, okay I will commit tomorrow and fix this. Thank you

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.cpp:27]:
```suggestion
    intensity->setDescription(QObject::tr("Crossfade the noise with the dry signal"));
```

File: src/effects/builtin/whitenoiseeffect.h
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.h:24]:
```suggestion
    CSAMPLE_GAIN previous_drywet;
```
no need to abbreviate this

File: src/effects/builtin/whitenoiseeffect.h
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.h:19]:
```suggestion
      : EffectState(bufferParameters),
        previous_drywet(0.0),
        gen(rs()) {
```
When switching between effects there can be a jarring jump in the audio. I think it may be because `previous_drywet` is uninitialised, so it could be any value.

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.cpp:75]:
```suggestion
if (enableState == EffectEnableState::Disabling) {
    gs.prev_drywet = 0;
} else {
    gs.prev_drywet = drywet;
}
```
This makes the effect ramp from dry when it is toggled on and off.

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.cpp:56]: remove, will be used below

**Comment by Be-ing** [comment]
This is more fun than I expected. I am glad you decided to work on it despite my scepticism. :) There are many potential chains to play with using this. I had fun with White Noise -> BitCrusher -> Moog Filter.

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.cpp:21]:
```suggestion
    pManifest->setDescription(QObject::tr("Creates active white noise sound"));
    pManifest->setEffectRampsFromDry(true);
```
The ramping code in the effect takes care of this if `previous_drywet` starts at 0 when the effect is loaded and disabled/enabled.

**Comment by uklotzde** [comment]
@katsar0v Please improve both the scope of your commits and your commit messages. Otherwise this will introduce a lot of noise into our commit history and complicate bisecting. This becomes unhandy within the scope of a large project.

If changes do not add real value from an outside perspective that could be summarized by a concise commit message then don't commit those changes individually. You could do this locally, but then please reorder and stash your commits into reasonable chunks before opening a PR.

I recommend to stash all the of commits of this branch before merging. The intermediate commits mostly document your personal learning process and don't add any value for the project.

See also: https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic
/commit.20message.20format

**Review by uklotzde — CHANGES_REQUESTED** [review]
I suggest to stash the intermediate commits. Some of them don't even build.

File: src/effects/builtin/whitenoiseeffect.cpp
Comment (Be-ing) [inline src/effects/builtin/whitenoiseeffect.cpp:21]:
```suggestion
    pManifest->setDescription(QObject::tr("Mix white noise with the input signal"));
```

**Comment by Be-ing** [comment]
The artifact when enabling/disabling and load the effect is gone now that `previous_drywet` is initialized to 0.

**Comment by Be-ing** [comment]
Sounds good to me! Just waiting for the CI builds to pass.

**Comment by Be-ing** [comment]
Travis macOS SCons build timed out but everything else passed. We should really get rid of SCons ASAP so we can save time on CI. The CMake builds go *way* faster with CCache.

**Comment by vespadj** [comment]
Dear developers, I would like to bring this topic to your attention:
https://github.com/mixxxdj/mixxx/issues/11868 improve usability of White Noise effect
Thank you

**Pull request merged**
